File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/shop/ShopViewModel.kt
package com.jworks.kanjiquest.android.ui.shop

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.UserSessionProvider
import com.jworks.kanjiquest.core.domain.model.CoinBalance
import com.jworks.kanjiquest.core.domain.model.PurchaseResult
import com.jworks.kanjiquest.core.domain.model.ShopCategory
import com.jworks.kanjiquest.core.domain.model.ShopItem
import com.jworks.kanjiquest.core.domain.repository.JCoinRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ShopUiState(
    val catalog: List<ShopItem> = emptyList(),
    val selectedCategory: ShopCategory? = null,
    val balance: CoinBalance = CoinBalance.empty(),
    val ownedContentIds: Set<String> = emptySet(),
    val isLoading: Boolean = true,
    val purchaseDialogItem: ShopItem? = null,
    val purchaseResult: PurchaseResult? = null,
    val featuredItem: ShopItem? = null,
    val hasRedeemedTrial: Boolean = false
) {
    val categories: List<ShopCategory>
        get() = catalog.map { it.category }.distinct()

    val filteredItems: List<ShopItem>
        get() = if (selectedCategory == null) catalog
                else catalog.filter { it.category == selectedCategory }
}

@HiltViewModel
class ShopViewModel @Inject constructor(
    private val jCoinRepository: JCoinRepository,
    private val userSessionProvider: UserSessionProvider
) : ViewModel() {

    private val _uiState = MutableStateFlow(ShopUiState())
    val uiState: StateFlow<ShopUiState> = _uiState.asStateFlow()

    init {
        loadShop()
        observeBalance()
    }

    private fun loadShop() {
        viewModelScope.launch {
            val userId = userSessionProvider.getUserId()
            val catalog = jCoinRepository.getShopCatalog()
            val owned = mutableSetOf<String>()
            for (category in ShopCategory.entries) {
                val unlocked = jCoinRepository.getUnlockedContent(userId, category.name.lowercase())
                owned.addAll(unlocked)
            }
            val featuredItem = catalog.find { it.id == "tutoringjay_trial" }
            val hasRedeemedTrial = featuredItem?.contentId != null && featuredItem.contentId in owned
            _uiState.value = _uiState.value.copy(
                catalog = catalog,
                ownedContentIds = owned,
                featuredItem = featuredItem,
                hasRedeemedTrial = hasRedeemedTrial,
                isLoading = false
            )
        }
    }

    private fun observeBalance() {
        viewModelScope.launch {
            val userId = userSessionProvider.getUserId()
            jCoinRepository.observeBalance(userId).collect { balance ->
                _uiState.value = _uiState.value.copy(balance = balance)
            }
        }
    }

    fun selectCategory(category: ShopCategory?) {
        _uiState.value = _uiState.value.copy(selectedCategory = category)
    }

    fun showPurchaseDialog(item: ShopItem) {
        _uiState.value = _uiState.value.copy(purchaseDialogItem = item, purchaseResult = null)
    }

    fun dismissPurchaseDialog() {
        _uiState.value = _uiState.value.copy(purchaseDialogItem = null, purchaseResult = null)
    }

    fun confirmPurchase(item: ShopItem) {
        viewModelScope.launch {
            val userId = userSessionProvider.getUserId()
            val result = jCoinRepository.purchaseItem(userId, item)
            _uiState.value = _uiState.value.copy(purchaseResult = result)

            if (result is PurchaseResult.Success) {
                refreshOwnedContent()
            }
        }
    }

    fun purchaseFeatured() {
        val featured = _uiState.value.featuredItem ?: return
        showPurchaseDialog(featured)
    }

    private suspend fun refreshOwnedContent() {
        val userId = userSessionProvider.getUserId()
        val owned = mutableSetOf<String>()
        for (category in ShopCategory.entries) {
            val unlocked = jCoinRepository.getUnlockedContent(userId, category.name.lowercase())
            owned.addAll(unlocked)
        }
        val featuredItem = _uiState.value.featuredItem
        val hasRedeemedTrial = featuredItem?.contentId != null && featuredItem.contentId in owned
        _uiState.value = _uiState.value.copy(
            ownedContentIds = owned,
            hasRedeemedTrial = hasRedeemedTrial
        )
    }
}

File: ios-app/KanjiQuest/Screens/Shop/ShopViewModel.swift
import Foundation
import SharedCore

@MainActor
final class ShopViewModel: ObservableObject {
    @Published var catalog: [ShopItem] = []
    @Published var selectedCategory: ShopCategory? = nil
    @Published var balance: Int64 = 0
    @Published var ownedContentIds: Set<String> = []
    @Published var isLoading = true
    @Published var purchaseDialogItem: ShopItem? = nil
    @Published var purchaseResult: PurchaseResult? = nil
    @Published var featuredItem: ShopItem? = nil
    @Published var hasRedeemedTrial = false

    var categories: [ShopCategory] {
        Array(Set(catalog.map { $0.category }))
    }

    var filteredItems: [ShopItem] {
        guard let cat = selectedCategory else { return catalog }
        return catalog.filter { $0.category == cat }
    }

    private var jCoinRepository: JCoinRepository?
    private var userSessionProvider: UserSessionProviderImpl?

    func load(container: AppContainer) {
        jCoinRepository = container.jCoinRepository
        userSessionProvider = container.userSessionProvider

        Task {
            let userId = userSessionProvider?.getUserId() ?? ""

            // Load catalog
            catalog = (try? await jCoinRepository?.getShopCatalog()) ?? []

            // Load balance
            balance = (try? await jCoinRepository?.getBalance(userId: userId))?.displayBalance ?? 0

            // Load owned items
            var owned = Set<String>()
            for cat in ShopCategory.allCases {
                let unlocked = (try? await jCoinRepository?.getUnlockedContent(userId: userId, category: cat.name.lowercased())) ?? []
                owned.formUnion(unlocked)
            }
            ownedContentIds = owned

            featuredItem = catalog.first { $0.id == "tutoringjay_trial" }
            hasRedeemedTrial = featuredItem?.contentId != nil && ownedContentIds.contains(featuredItem?.contentId ?? "")

            isLoading = false
        }
    }

    func selectCategory(_ category: ShopCategory?) {
        selectedCategory = category
    }

    func showPurchaseDialog(_ item: ShopItem) {
        purchaseDialogItem = item
        purchaseResult = nil
    }

    func dismissPurchaseDialog() {
        purchaseDialogItem = nil
        purchaseResult = nil
    }

    func confirmPurchase(_ item: ShopItem) {
        Task {
            let userId = userSessionProvider?.getUserId() ?? ""
            let result = try? await jCoinRepository?.purchaseItem(userId: userId, item: item)
            purchaseResult = result

            if case .success = result {
                // Refresh owned
                var owned = Set<String>()
                for cat in ShopCategory.allCases {
                    let unlocked = (try? await jCoinRepository?.getUnlockedContent(userId: userId, category: cat.name.lowercased())) ?? []
                    owned.formUnion(unlocked)
                }
                ownedContentIds = owned
                hasRedeemedTrial = featuredItem?.contentId != nil && owned.contains(featuredItem?.contentId ?? "")
            }
        }
    }

    func purchaseFeatured() {
        guard let item = featuredItem else { return }
        showPurchaseDialog(item)
    }
}
